### ISSUE TITLE
Design data filtering strategy for the frontend

### ISSUE THREAD

**Issue by Anakin100100** [po8klasie/po8klasie-rails-api#7]
A comprehensive strategy of what filters on the data will be required is needed. This specification needs to match with the needs of the frontend application. 
The categories of filters that i can think of right now are:
 - Filters based on location (selecting the area needs to be quite granual) inside a region, area or county
 - Filters based on geographical coordinates (user supplied or predefined)
 - Filters based on educational statistics (not very useful for most of the schools)

**Comment by micorix** [comment]
### Filtering
 Below I listed filters that are crucial for deploying initial version of our site:

* query (string)
  *  Basically school name with some fuzzy search for now but maybe later we'd query multiple columns (eg. school name & school name shortened or sth like that)
* is school public or not (bool/multiple choice)
  * maybe it shouldn't even be boolean, maybe we should make "szkoła społeczna"?
* school type (multiple choice)
  * liceum ogólnokształcące, technikum, szkoła branżowa {I,II} stopnia, etc.
* district (multiple choice)
* extended subjects (multiple choice)

We decided to use OR operator for multiple choice questions.

Note: This list may change in the future (we will likely add more filters, not remove existing ones)

### Sorting
When it comes to sorting I think we should just stick to alphabetical order by default. 

Another thing is that we really don't want to show users school pages without any data, so I'd figure sth out to prevent that.
For example: if we don't have info about extended subjects for some school, we can "move" that school to the end of search results. 

Since moving schools freely back and forth would not be intuitive for the user, maybe we should have some  "show schools with X data first" switch on the frontend (enabled by default) and then would send some param ( eg. `?prioritizeSchoolsBasedOnData=true`) to the backend. 

Why is it important? Bc users who see empty pages right after entering our app, may think that sth broke/is wrong with our app. 

But ofc implementing that shouldn't be our priority if all the schools have similar data, like in case of Gdynia. We should implement that when we start preparing for project rollout in Warsaw. Most likely we won't have all needed data.

### Filtering by area

Ofc we should implement filtering by area (city) we support, but idk if it needs to be THAT granular. Maybe string with areaName would suffice?

**Comment by Anakin100100** [comment]
@micorix thanks that should be enough to get us going

**Comment by Anakin100100** [comment]
@wojtodzio There are quite a lot of filters there at once so we would have to make one  sql query that would eliminate the most records (fe. area) and do the rest of the data manipulation on the server. I don't think this would be very performant and easy to maintain. Another option would be to write the whole search query as one PL/SQL function and execute the search entirely on the database. This would be much more effecient on it's own and would allow us to utilise query caching (people are likely to make the same queries accross many sessions there are several different groups of students that use similar queries in an area) very effeciently. What do you think about both possible implementations?

**Comment by wojtodzio** [comment]
I agree manipulating data on the server won't be performant, but SQL functions are tough to maintain.
Rails' Active Record is lazily evaluated, which allows for building complex queries in an easy to maintain and performant way. E.g., the following function builds query conditionally and fires only one DB query: 
```ruby
def find_users(ids, email_domain, admins_only)
  users = User.all
  users = users.where(id: ids) if ids.present?
  users = users.where('email ILIKE ?', "%#{email_domain}%") if email_domain.present?
  users = query.where(admin: true) if admins_only?
  users
end
```
(this is just a pseudo code to illustrate the functionality, I didn't test this particular function)

I'd also suggest looking for gems that can help you with that, as filtering is a pretty common thing to implement. A few ones that might be useful: https://github.com/heartcombo/has_scope https://github.com/Casecommons/pg_search https://github.com/activerecord-hackery/ransack

**Comment by Anakin100100** [comment]
@wojtodzio The has_scope doesn't seem to be granular enough and is hard to implement easily for an api. The pg_serach seems to perfectly fulfill our need for full text search so that we don't need elasticsearch. The advanced mode in ransack provides very granular control over the search logic which we are going to need. [this](https://stackoverflow.com/questions/31792183/how-can-ransack-use-pg-search) Stack Overflow post suggests that we could easily use both of these together to create one query. Do you think that using pg_search for text search and ransack for detailed filtering would be a good idea?

**Comment by wojtodzio** [comment]
Actually, now that I think about it, I haven't used ransack in an API-only app yet, and I'm not sure how convenient it will be to work with, but it may be worth giving it a shot. You should definitely be able to combine it with pg_search, as they both operate on AR scopes.

**Comment by Anakin100100** [comment]
The ransack gem has extremely conveniant native way of passing search params from the request to the querry. Let's settle on the combination of both for now.

**Comment by Anakin100100** [comment]
I don't think that we will have too much use for ransack because it is used almost exclusively at the frontend and isn't of too much use for the backend. @micorix how are we supposed to get the extended subjects that a school offers? For now we don't have a method of extracting that information.

**Comment by micorix** [comment]
It will be included in data from Gdynia soon (ig in the coming week)

**Comment by Anakin100100** [comment]
How about Warsaw?

**Comment by micorix** [comment]
I expect we have it in next month (if we're lucky)
